### ISSUE TITLE
`gem.bat` on Windows inaccurate

### ISSUE THREAD

**Issue by ccmywish** [rubygems/rubygems#6483]
### Describe the problem as clearly as you can

I'm on Windows 11 using `Ruby 3.2.0-1` built from RubyInstaller2. 

After I update `gem` from `3.4.2` to the latest `3.4.8`, I got a new file `gem.bat`. I'm very sure THIS IS A NEW FILE generated after updating, because I'm developing [rbenv for Windows](https://github.com/ccmywish/rbenv-for-windows), and it relies on the bin stub file. Before updating, there only exists a `gem.cmd` as a bin stub file. After updating, I got `gem.bat` so that from then when I invoke `gem`, the script `gem.bat` is called rather than the old `gem.cmd`.

However, in this `gem.bat`, inaccurate version of `ruby.exe` is called: 

```batch
# This is C:\Ruby-on-Windows\3.2.0-1\bin\gem.bat
  @ECHO OFF
  IF NOT "%~f0" == "~f0" GOTO :WinNT
  @"ruby.exe" "C:/Ruby-on-Windows/3.2.0-1/bin/gem" %1 %2 %3 %4 %5 %6 %7 %8 %9
  GOTO :EOF
  :WinNT
  @"ruby.exe" "%~dpn0" %*
```
You can see that, the `@"ruby.exe"` is going to invoke the `ruby.exe` in `PATH` environment variable. Instead, we should invoke the `ruby.exe` next to the `gem.bat` in `C:\Ruby-on-Windows\3.2.0-1\bin\`, because only in this way, can we run the same version of `ruby.exe` with `gem.bat`. 

So, we should change from `@"ruby.exe"` to `@"%~dp0ruby.exe"`. 

**In fact, I don't know why this strange `gem.bat` got generated, because it's very different to other commands generated in `xxx.bat`.**

```bat
# Other normal gem commands in xxx.bat
@ECHO OFF
@"%~dp0ruby.exe" "%~dpn0" %*
``` 

**If we can omit generating this `gem.bat`, things will work too without above changes.**

<br>

### Did you try upgrading RubyGems?

Yes, it's the issue of the latest RubyGems.

### Post steps to reproduce the problem

`gem update --system --pre`

**Comment by MSP-Greg** [comment]
Given options for building Ruby (suffixes), and also how `Gem.ruby` handles a path with spaces, the following would be best:
```
@ECHO OFF
@"%~dp0#{File.basename(Gem.ruby).chomp('"')}" -x "%~dpn0" %*
```

As to the `IF NOT "%~f0" == "~f0" GOTO :WinNT` line, I'm not even sure that ever gets 'hit' with current Windows versions.  Or, on my system, it always executes the last line by jumping to `:WinNT`...

To test this, I copied a master build (3.3.0) to `C:\Program Files\ruby-ucrt`, then switched my path to use Ruby 3.2.  Then, from `C:\Program Files\ruby-ucrt`, I ran `bin/gem env`, and got the following:
```
  - RUBYGEMS VERSION: 3.5.0.dev
  - RUBY VERSION: 3.3.0 (2023-03-15 patchlevel -1) [x64-mingw-ucrt]
```

If I run `gem env` from the same folder, I get:

```
  - RUBYGEMS VERSION: 3.4.8
  - RUBY VERSION: 3.2.1 (2023-02-08 patchlevel 31) [x64-mingw-ucrt]
```

**Comment by deivid-rodriguez** [comment]
Did you ever figure this out @ccmywish?

**Comment by ccmywish** [comment]
Hi @deivid-rodriguez 

# Before updating `RubyGems`:
![image](https://github.com/rubygems/rubygems/assets/63459097/e2f9caaf-7268-4fd2-aa8f-81d043a9509f)

**All other gems' executables are `.bat`**

# After updating `RubyGems`:
Look！A new `gem.bat`
![image](https://github.com/rubygems/rubygems/assets/63459097/ca5884a7-1aab-4c6d-8571-2fb931304827)

The related code is here:
https://github.com/rubygems/rubygems/blob/master/lib/rubygems/commands/setup_command.rb#L279-L287

As @MSP-Greg pointed out, the part before `:WinRT` is outdated. 

**In fact, I saw it last month from my `Perl` on Windows. The code below is generated by `pl2bat.bat` from Perl's distribution.**

![image](https://github.com/rubygems/rubygems/assets/63459097/6a25c15b-7627-4d72-b515-ea56bf5ec7d6)

# Changes to be made
So, I think there should be two changes:

1. I think we can and we should delete the `non-:WinRT` part totally now. All other gem executables for example `jekyll.bat` don't have these contens at all.

2. Another change should be made is that, also according to @MSP-Greg, add a `%~dp0` before `#{File.basename(Gem.ruby).chomp('"')}`

**Comment by deivid-rodriguez** [comment]
Nice! Would you be willing to create a PR to fix these things?

**Comment by ccmywish** [comment]
Sure I'd be happy to do that. I've created a PR and add the co-author info. #7785

**Comment by deivid-rodriguez** [comment]
Closed by #7785, thank you!
